### PR TITLE
Add lazy memory operation mode and config WIP

### PR DESCRIPTION
**Pull request by amitmarkel**
This is a first version that allows lazy operation. Next steps may be:

1. Change multimet's design to allow preloading selective basins-wise or samples-wise every epoc (and in parallel by cpu, can load for the next epoc), dropping loaded data for each epoc.
2. Allow the validation set to always be non-lazy via a flag.

---

Benchmark difference for handoff on CPU.

batch size=2
3 basins
train 10/2012-09/2017
validate 10/2017-09/2022

Non-lazy (eager; in-memory):
```
[INFO] 18:22:46.914 (basetrainer.py:train_and_validate) -- learning rate is [0.001]
# Epoch 1: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:05<00:00,  1.12s/it, Loss: 0.9693]
[INFO] 18:22:52.526 (basetrainer.py:train_and_validate) -- Epoch 1 average loss: avg_loss: 0.30864, avg_forecast_overlap: 0.48499, avg_total_loss: 0.79363
[INFO] 18:22:52.535 (basetrainer.py:train_and_validate) -- learning rate is [0.001]
# Epoch 2: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:01<00:00,  4.43it/s, Loss: 0.2500]
[INFO] 18:22:53.665 (basetrainer.py:train_and_validate) -- Epoch 2 average loss: avg_learning_rate: 0.00100, avg_loss: 2.23824, avg_forecast_overlap: 0.20897, avg_total_loss: 2.44721
Metrics for 1D are calculated over last 1 elements only. Ignoring 7 predictions per sequence.
# Validation: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:32<00:00, 16.37s/it]
[INFO] 18:23:26.417 (basetrainer.py:train_and_validate) -- Epoch 2 average validation loss: 1.25240 -- Median validation metrics: avg_loss: 1.10535, avg_forecast_overlap: 0.14808, NSE: -4.92259
[INFO] 18:23:26.417 (basetrainer.py:train_and_validate) -- learning rate is [0.001]
# Epoch 3: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:01<00:00,  4.11it/s, Loss: 7.0540]
[INFO] 18:23:27.636 (basetrainer.py:train_and_validate) -- Epoch 3 average loss: avg_learning_rate: 0.00100, avg_loss: 3.54916, avg_forecast_overlap: 0.23636, avg_total_loss: 3.78552
[INFO] 18:23:27.645 (basetrainer.py:train_and_validate) -- learning rate is [0.001]
# Epoch 4: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:01<00:00,  3.84it/s, Loss: 0.4908]
[INFO] 18:23:28.951 (basetrainer.py:train_and_validate) -- Epoch 4 average loss: avg_learning_rate: 0.00100, avg_loss: 0.55217, avg_forecast_overlap: 0.10159, avg_total_loss: 0.65376
Metrics for 1D are calculated over last 1 elements only. Ignoring 7 predictions per sequence.
# Validation: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:32<00:00, 16.48s/it]
[INFO] 18:24:01.933 (basetrainer.py:train_and_validate) -- Epoch 4 average validation loss: 0.79873 -- Median validation metrics: avg_loss: 0.73917, avg_forecast_overlap: 0.06832, NSE: -0.64245
[INFO] 18:24:01.933 (basetrainer.py:train_and_validate) -- learning rate is [0.001]
# Epoch 5: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:01<00:00,  4.38it/s, Loss: 0.3610]
[INFO] 18:24:03.770 (basetrainer.py:train_and_validate) -- Epoch 5 average loss: avg_learning_rate: 0.00100, avg_loss: 0.18490, avg_forecast_overlap: 0.10309, avg_total_loss: 0.28799
[INFO] 18:24:03.860 (basetrainer.py:train_and_validate) -- learning rate is [0.001]
# Epoch 6: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:01<00:00,  4.39it/s, Loss: 0.3903]
[INFO] 18:24:04.228 (basetrainer.py:train_and_validate) -- Epoch 6 average loss: avg_learning_rate: 0.00100, avg_loss: 0.28731, avg_forecast_overlap: 0.11051, avg_total_loss: 0.39782
Metrics for 1D are calculated over last 1 elements only. Ignoring 7 predictions per sequence.
# Validation: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:32<00:00, 16.09s/it]
[INFO] 18:24:36.431 (basetrainer.py:train_and_validate) -- Epoch 6 average validation loss: 0.99952 -- Median validation metrics: avg_loss: 0.94329, avg_forecast_overlap: 0.05728, NSE: -0.13803
[INFO] 18:24:36.431 (basetrainer.py:train_and_validate) -- learning rate is [0.001]
# Epoch 7: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:01<00:00,  4.63it/s, Loss: 0.5010]
[INFO] 18:24:37.514 (basetrainer.py:train_and_validate) -- Epoch 7 average loss: avg_learning_rate: 0.00100, avg_loss: 0.21160, avg_forecast_overlap: 0.14271, avg_total_loss: 0.35431
[INFO] 18:24:37.523 (basetrainer.py:train_and_validate) -- learning rate is [0.001]
# Epoch 8: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:01<00:00,  4.36it/s, Loss: 0.2507]
```

Lazy:
```
# Epoch 1: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:26<00:00,  5.29s/it, Loss: 0.5078]
[INFO] 18:13:48.958 (basetrainer.py:train_and_validate) -- Epoch 1 average loss: avg_loss: 0.35290, avg_forecast_overlap: 0.29720, avg_total_loss: 0.65010
# Validation:  33%|██████████████████████████████████████████████████████████████████████████                                                                                                                                                    | 1/3 [04:26<08:53, 266.58s/it]
```

Lazy (batch size=32)
```
# Epoch 2: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 6/6 [00:46<00:00,  7.82s/it, Loss: 0.6658]
```

### CODE REVIEW

File: googlehydrology/datasetzoo/multimet.py [hunk 1]
```diff
@@ -360,15 +367,18 @@ def _extract_hindcasts(self, item: int) -> dict[str, np.ndarray]:
...
 
-        return {name: np.expand_dims(feature, -1) for name, feature in features.items()}
+        return {name: self._comp.expand_dims(feature, -1) for name, feature in features.items()}
         # TODO (future) :: This adds a dimension to many features, as required by some models.
         # There is no need for this except that it is how basedataset works, and everything else expects
         # the trailing dim. Remove this dependency in the future.
 
     def _extract_forecasts(self, item: int) -> dict[str, np.ndarray]:
+        forecast_indexer = self._sample_index[item].copy()
+        forecast_indexer["lead_time"] = slice(MULTIMET_MINIMUM_LEAD_TIME, None)
```
Comment (omrishefi) [inline googlehydrology/datasetzoo/multimet.py]: Can it be slice(None,None)?

File: googlehydrology/datasetzoo/multimet.py [hunk 2]
```diff
@@ -681,14 +692,15 @@ def collate_fn(
...
-    data: xr.DataArray, indexers: dict[Hashable, int | range]
+    data: xr.DataArray, indexers: dict[Hashable, int | range | slice]
 ) -> np.ndarray | np.float32:
     """Returns the values in array according to dims given by indexers.
 
     This function replaces uses of `isel` with data and indexers.
     """
-    locators = (indexers[dim] if dim in indexers else slice(None) for dim in data.dims)
-    return data.data[tuple(locators)]
+    locs = (indexers[dim] if dim in indexers else slice(None) for dim in data.dims)
```
Comment (omrishefi) [inline googlehydrology/datasetzoo/multimet.py:850]: Can we load data.dims without loading all the data?

File: googlehydrology/utils/config.py [hunk 3]
```diff
@@ -297,6 +311,20 @@ def _read_and_parse_config(yml_path: Path):
...
         return self._cfg.get("detect_anomaly", False)
 
+    @property
+    def lazy_memory(self) -> LazyMemory:
+        data = self._cfg.get('lazy_memory', {})
+        data = pydantic.TypeAdapter(LazyMemory).validate_python(data)
+
+        if data.enabled and not self._cache:
+            LOGGER.info('Lazy Memory on with %d cache bytes', data.cache_bytes)
+            self._cache = dask.cache.Cache(
```
Comment (omrishefi) [inline googlehydrology/utils/config.py:321]: We shouldn't add cache inside Config layer

Comment (amitmarkel) [reply googlehydrology/utils/config.py:321]: This will be changed.

File: googlehydrology/datasetzoo/multimet.py (hunk 2 above)
Comment (amitmarkel) [reply googlehydrology/datasetzoo/multimet.py:850]: IIRC this was only to support slices

dims are available when not doing a compute() yet (without loading all the data)

File: googlehydrology/datasetzoo/multimet.py (hunk 1 above)
Comment (amitmarkel) [reply googlehydrology/datasetzoo/multimet.py]: Redone this as a new PR so this isn't needed anymore per the new arch.
